BrainPortal/app/controllers/data_providers_controller.rb

```diff
@@ -115,25 +115,31 @@ def new_personal #:nodoc:
     @provider         = UserkeyFlatDirSshDataProvider.new( :user_id   => current_user.id,
                                                            :group_id  => provider_group_id,
                                                            :online    => true,
-                                                           :read_only => false
+                                                           :read_only => false,
                                                          )
     @groups           = current_user.assignable_groups
   end
 
-  # create by normal user,  only UserkeyFlatDirSshDataProvider
+  # Can be create by normal user,
+  # only UserkeyFlatDirSshDataProvider, S3FlatDataProvider, S3MultiLevelDataProvider
   def create_personal
-    normal_params = params.require_as_params(:data_provider)
-                          .permit(:name, :description, :group_id,
-                                  :remote_user, :remote_host,
-                                  :remote_port, :remote_dir
-                                 )
-    group_id = normal_params[:group_id]
-    current_user.assignable_group_ids.find(group_id) # ensure assignable, not sure need check visibility etc more
-    @provider         = UserkeyFlatDirSshDataProvider.new(normal_params)
-    @provider.user_id = current_user.id # prevent creation of dp on behalf of other users
+    dp_type   = (params[:data_provider][:type] || "UserkeyFlatDirSshDataProvider").constantize
+
+    @provider = dp_type.new(base_provider_params)
```

Comment:
Why extract the DP type and not let the STI code handle the tpying directly from the params?

```ruby
@provider = DataProvider.new(base_provider_params)
```

would work even better given `:type` is already one of the attributes returned by `common_params_list()`